### ISSUE TITLE
rustdoc: Can't use anchors in links to primitive types

### ISSUE THREAD

**Issue by taylordotfish** [rust-lang/rust#83083]
rustdoc doesn't allow anchors in links to primitive types. For example, given this code:

```rust
//! [test](str#examples)
```

`cargo doc` produces this warning:

```
warning: `str#examples` contains an anchor, but links to builtin types are already anchored
 --> src/lib.rs:1:12
  |
1 | //! [test](str#examples)
  |            ^^^^^^^^^^^^ contains invalid anchor
  |
  = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
```

and the resulting link doesn't point to any part of `str`'s documentation, but is rather treated as a normal URL relative to the current page.

The warning claims that “links to builtin types are already anchored”, but this does not appear to be correct. For example, this code:

```rust
//! [test](str)
```

produces a link to <https://doc.rust-lang.org/nightly/std/primitive.str.html>, which does not contain an anchor.

### Meta

```
rustc 1.52.0-nightly (b3e19a221 2021-03-12)
binary: rustc
commit-hash: b3e19a221e63dcffdef87e12eadf1f36a8b90295
commit-date: 2021-03-12
host: powerpc64le-unknown-linux-gnu
release: 1.52.0-nightly
LLVM version: 12.0.0
```

**Comment by jyn514** [comment]
This was first made an error in 454d13badf6a5d423ba1cc56d106137752db719c. At that point there was a comment saying "TODO: give an error", which was first added in 09293bec925241782e5d173292fad68735d6961a.

@GuillaumeGomez do you remember why you made this an error? https://github.com/rust-lang/rust/pull/66675 was released in 1.41.0, which still uses the `primitive.str.html` URL scheme.

**Comment by GuillaumeGomez** [comment]
I think I had in mind that the anchors weren't stable and that it wasn't "safe" to use anchors on primitive for this reason. Not sure if it makes much sense though considering we allow it for foreign types...

**Comment by jyn514** [comment]
Oh, I found why this restriction is here: https://github.com/rust-lang/rust/blob/acca818928654807ed3bc1ce0e97df118f8716c8/src/librustdoc/passes/collect_intra_doc_links.rs#L238-L243
I'm ambivalent whether this should be considered a bug or a feature request; happy for anyone to try and fix it but it won't be trivial. The fragment is used here: https://github.com/rust-lang/rust/blob/acca818928654807ed3bc1ce0e97df118f8716c8/src/librustdoc/passes/collect_intra_doc_links.rs#L1252 and later here: https://github.com/rust-lang/rust/blob/acca818928654807ed3bc1ce0e97df118f8716c8/src/librustdoc/clean/types.rs#L873-L902

**Comment by jyn514** [comment]
This would be a good use case for my 'fake DefId' idea: https://rust-lang.zulipchat.com/#narrow/stream/266220-rustdoc/topic/getting.20rid.20of.20cell.20and.20refcell.20usage/near/225222711
